### ISSUE TITLE
Operation not permitted when adding gnome remote

### ISSUE THREAD

**Issue by Powersource** [flatpak/flatpak#127]
I was following the steps listed here http://flatpak.org/index.html#users but flatpak wouldn't let me add the first nor second remote. The steps mention sudo on other commands but not these so I'm guessing this is some issue with gpg? Running arch and flatpak v. 0.6.5-1.

```
me@pc[~/dl]% wget https://sdk.gnome.org/keys/gnome-sdk.gpg
--2016-06-18 23:32:34--  https://sdk.gnome.org/keys/gnome-sdk.gpg
Resolving sdk.gnome.org (sdk.gnome.org)... 209.132.180.169
Connecting to sdk.gnome.org (sdk.gnome.org)|209.132.180.169|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 629
Saving to: ‘gnome-sdk.gpg’

gnome-sdk.gpg            100%[=================================>]     629  --.-KB/s    in 0s      

2016-06-18 23:32:35 (62.6 MB/s) - ‘gnome-sdk.gpg’ saved [629/629]

me@pc[~/dl]% flatpak remote-add --gpg-import=gnome-sdk.gpg gnome https://sdk.gnome.org/repo/
error: GDBus.Error:org.gtk.GDBus.UnmappedGError.Quark._g_2dio_2derror_2dquark.Code14: Operation not permitted
me@pc[~/dl]% ls *.gpg
gnome-sdk.gpg
me@pc[~/dl]%
```

**Comment by alexlarsson** [comment]
I believe this is some sort of packaging issue. What happens here is that the client (flatpak remote-add) talks to the flatpak-system-helper running as root, asking it to create a remote. The system helper then uses policykit to verify that the client is allowed to do this. Adding a remote requires admin privs, so it should then ask for your password, and only if that fails it will return "no permission". Did you get a password dialog? If not, maybe the policykit configuration is somehow wrong in the arch package?

@heftig: Any ideas?

**Comment by heftig** [comment]
If this is a deny from polkit, shouldn't it return `org.freedesktop.DBus.Error.AccessDenied` instead of an unmapped `GIOError`?

**Comment by heftig** [comment]
> If this is a deny from polkit, shouldn't it return org.freedesktop.DBus.Error.AccessDenied instead of an unmapped GIOError?

Nope, it returns a `GIOError` ([system-helper.c#L696](https://github.com/flatpak/flatpak/blob/master/system-helper/flatpak-system-helper.c#L696)).

Maybe the user isn't in the `wheel` group?

**Comment by alexlarsson** [comment]
@heftig That should only affect the install/update policy. Adding a remote is auth_admin, so should ask for the full root passwork if the user is not wheel. No?

**Comment by heftig** [comment]
AFAIK Polkit for `auth_admin`, will enumerate the `wheel` group and allow the user to pick a member to authenticate as. Of course, this can only work if the `wheel` group has at least one member.

**Comment by Powersource** [comment]
I am in wheel and it didn't ask me for a password.

**Comment by heftig** [comment]
Strange. Does `pkexec sh` work?

**Comment by Powersource** [comment]
Huh, I did get the usual dialog I get when using systemctl but while systemctl seems to still work, this breaks.

```me@pc[~/dl]% pkexec sh
==== AUTHENTICATING FOR org.freedesktop.policykit.exec ===
Authentication is needed to run `/usr/bin/sh' as the super user
Authenticating as: me
Password: 
polkit-agent-helper-1: error response to PolicyKit daemon: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: No session for cookie
==== AUTHENTICATION FAILED ===
Error executing command as another user: Not authorized

This incident has been reported.
me@pc[~/dl]%```

Closest thing I found was this https://bbs.archlinux.org/viewtopic.php?id=210460 but that sounds specific to that other program?

**Comment by Powersource** [comment]
Or wait, this sounds a lot more relevant. I'm guessing we can close this issue (what's your policy?) and I'll go nag them instead. https://bugs.freedesktop.org/show_bug.cgi?id=90837

**Comment by alexlarsson** [comment]
closing this for the fd.o bug then

**Comment by dmerejkowsky** [comment]
Hum. Got the same error, but it disappeared when I switch from running fluxbox with startx from a tty, to a gnome session started by gdm ...

Maybe it was simply because some daemon was not running?

**Comment by dmerejkowsky** [comment]
Hum. Got the same error, but it disappeared when I switch from running fluxbox with startx from a tty, to a gnome session started by gdm ...

Maybe it was simply because some daemon was not running?

**Comment by Powersource** [comment]
Thanks @dmerejkowsky you got me to look at this just a bit more and realize my previous conclusion was far off. Like you I'm just running a WM (i3wm) (with some xfce daemons slapped on top) which breaks stuff from time to time. I have a shitton of tabs open atm. but I think [this reddit thread](https://www.reddit.com/r/debian/comments/4h33rs/pkexec_authentication_failure/) summarizes it well. Basically for some reason using the built in polkit agent `pkttyagent` doesn't work "in graphical environments" (not that I could get it to work either in a tty).

What I did to solve it (albeit hackishly) was:

1. Install polkit-gnome (at least in arch).
2. Run `/usr/lib/polkit-gnome/polkit-gnome-authentication-agent-1` on boot (full path is needed. The package gives you an unusable for me .desktop file but that pointed me to the binary).

This is normally handled by a session manager provided by your DE. I still don't understand polkit fully but I see no reason for why the issues I ran into should happen. At least they have a problem with error messages.
